### DISCUSSION TITLE
retrofit IllegalArgumentException - minifyenabled true

### DISCUSSION THREAD

**Discussion by markusGF** [square/retrofit#4035] [General]
Hello,

i've searched nearly two days for an answer but nothing helps.
I want to implement retrofit in my project because of the deprecated AsyncTask function.

Now i have the following problem:
I have to use minifyenabled = true in my release, there is no option to disable it.
But when I enable this feature I get always the following Exception:

java.lang.IllegalArgumentException: Unable to create call adapter for interface retrofit2.Call for method TCWebserviceApi.logout

Thats my Configurations/Code:

------------------------------------------------------------------------------------------------------
Gradle.Bundle:

implementation 'com.squareup.retrofit2:retrofit:2.9.0'
implementation 'com.squareup.retrofit2:converter-scalars:2.9.0'

------------------------------------------------------------------------------------------------------
My Interface:

    public interface TCWebserviceApi {
        // Request method and URL specified in the annotation

        @POST("logout")
        Call<String> logout(@Query("session_id") String session_id);
    }

------------------------------------------------------------------------------------------------------
My RestClient:

    public RestClient(Context context) {
        try {
            //Init Database
            mDb = new DatabaseHandler(context);

            //Beziehe Session Objekt
            refreshSession();

            this.BASE_URL = SOME_URL_STUFF;

            //Erstelle Retrofit Instanz
            mRetrofit = new Retrofit.Builder()
                    .baseUrl(BASE_URL)
                    .addConverterFactory(ScalarsConverterFactory.create())
                    .build();

            //Init TCWebserviceApi
            mWebserviceApi = mRetrofit.create(TCWebserviceApi.class);

        }catch(Exception e) {
            Log.e(TAG, "RestClient: ", e);
        }
    }
    
    //The logout function...
    public void logout(OnRestCall callback) {

            //Starte POST Request
            try {
                Call<String> call = mWebserviceApi.logout(mSession.getSession_Id());
                call.enqueue(new Callback<>() {
                    @Override
                    public void onResponse(Call<String> call, Response<String> response) {
                        //String result = response.body();

                        if (callback != null) {
                            if (response.isSuccessful()) {
                                callback.onSuccess("", response.code());
                            } else {
                                callback.onFailure("", response.code());
                            }
                        }
                    }

                    @Override
                    public void onFailure(Call<String> call, Throwable t) {
                        //String error = t.toString();
                        if (callback != null) {
                            callback.onException(t);
                        }
                    }
                });
            } catch (Exception e) {
                Log.e(TAG, "logout: ", e);
            }
    }
    
    

I dont use any converters, need only string as result to do further work...
Gradle Version 8.5

 I've tried to add the retrofit2.pro, add some rules to the proguard-rules.pro but nothing helps.
 
 How can i solve the problem?

**Comment by andymach** [comment]
{can i have your proguard rule file} when u minifyenabled = true u have to add Proguard rules and u have to use ignore case inside that for your retrofit package and java file

**Comment by markusGF** [comment]
Hello Andymarch,
thanks for your response.

Here are my proguard rules, there are only rules related to retrofit.
Dont wonder, i've added comment slashes to the comments, otherwise all comments shown here as an headline.

//#-------------------------------------------------------------------------------------------------------------------------

//# Retrofit does reflection on generic parameters. InnerClasses is required to use Signature and
//# EnclosingMethod is required to use InnerClasses.
-keepattributes Signature, InnerClasses, EnclosingMethod

//# Retrofit does reflection on method and parameter annotations.
-keepattributes RuntimeVisibleAnnotations, RuntimeVisibleParameterAnnotations

//# Keep annotation default values (e.g., retrofit2.http.Field.encoded).
-keepattributes AnnotationDefault

//# Retain service method parameters when optimizing.
-keepclassmembers,allowshrinking,allowobfuscation interface * {
    @retrofit2.http.* <methods>;
}

//# Ignore annotation used for build tooling.
-dontwarn org.codehaus.mojo.animal_sniffer.IgnoreJRERequirement

//# Ignore JSR 305 annotations for embedding nullability information.
-dontwarn javax.annotation.**

//# Guarded by a NoClassDefFoundError try/catch and only used when on the classpath.
-dontwarn kotlin.Unit

//# Top-level functions that can only be used by Kotlin.
-dontwarn retrofit2.KotlinExtensions
-dontwarn retrofit2.KotlinExtensions$*

//# With R8 full mode, it sees no subtypes of Retrofit interfaces since they are created with a Proxy
//# and replaces all potential values with null. Explicitly keeping the interfaces prevents this.
-if interface * { @retrofit2.http.* <methods>; }
-keep,allowobfuscation interface <1>

//# Keep inherited services.
-if interface * { @retrofit2.http.* <methods>; }
-keep,allowobfuscation interface * extends <1>

//# With R8 full mode generic signatures are stripped for classes that are not
//# kept. Suspend functions are wrapped in continuations where the type argument
//# is used.
-keep,allowobfuscation,allowshrinking class kotlin.coroutines.Continuation

//# R8 full mode strips generic signatures from return types if not kept.
-if interface * { @retrofit2.http.* public *** *(...); }
-keep,allowoptimization,allowshrinking,allowobfuscation class <3>

//# With R8 full mode generic signatures are stripped for classes that are not kept.
-keep,allowobfuscation,allowshrinking class retrofit2.Response

**Reply by Endeavour233** [reply]
Can you reproduce the issue now? I tried to reproduce it but failed. If so, could you please print out the cause of the caught Exception as well. Thanks.
```
 catch (Exception e) {
            Log.e(TAG, "logout: ", e);
            Log.e(TAG, "cause: ", e.cause);
        }
```
